cmd/testperformance.go

```diff
@@ -187,102 +217,241 @@ func testPerformance(ctx context.Context, queuedTests []testCaseName, allTests m
-		z.Any("loops", diskWriteLoops))
-
-	if conf.DiskWriteMB <= 3072 {
-		log.Warn(ctx, "File size used for tests of 3072MB or lower impacts the measured performance", nil)
+func fioCommand(ctx context.Context, filename string, fileSize int, blocksize int, operation string) ([]byte, error) {
+	//nolint:gosec
+	cmd, err := exec.CommandContext(ctx, "fio",
+		"--name=fioTest",
+		fmt.Sprintf("--filename=%v", filename),
+		fmt.Sprintf("--size=%vMb", fileSize),
+		fmt.Sprintf("--blocksize=%vk", blocksize),
+		fmt.Sprintf("--numjobs=%v", diskOpsNumOfJobs),
+		fmt.Sprintf("--rw=%v", operation),
+		"--direct=1",
+		"--runtime=60s",
+		"--group_reporting",
+		"--output-format=json",
+	).Output()
+	if err != nil {
+		return nil, errors.Wrap(err, "exec fio command")
 	}
 
+	return cmd, nil
+}
+
+func setupDiskOperation(ctx context.Context, conf *testPerformanceConfig) (int, string, error) {
 	var stat unix.Statfs_t
-	wd, _ := os.UserHomeDir()
-	err := unix.Statfs(wd, &stat)
+	var dir string
+	var err error
+	if conf.DiskIOTestFileDir == "" {
+		dir, err = os.UserHomeDir()
+		if err != nil {
+			return 0, "", errors.Wrap(err, "get user home dir")
+		}
+	} else {
+		dir = conf.DiskIOTestFileDir
+	}
+	err = unix.Statfs(dir, &stat)
 	if err != nil {
-		return failedTestResult(testRes, err)
+		return 0, "", errors.Wrap(err, "unix stat free storage")
 	}
-	// Available blocks * size per block = available space in bytes; remove 20% for safety
+
+	// Available blocks * size per block = available space in bytes; convert to MBs; use only 80% for safety
 	availableMB := int(stat.Bavail*uint64(stat.Bsize)/1024/1024) / 5 * 4
-	actualDiskWriteMB := conf.DiskWriteMB
+	diskWriteMB := 8192
 
-	for availableMB < actualDiskWriteMB {
-		log.Warn(ctx, fmt.Sprintf("Insufficient available disk space of %vMB, reducing the test size to %vMB. Note that this might result in slower write speed", availableMB, actualDiskWriteMB), nil)
-		actualDiskWriteMB /= actualDiskWriteMB
-		if actualDiskWriteMB == 0 {
-			return failedTestResult(testRes, errors.New("insufficient available disk space", z.Str("available_space", strconv.Itoa(availableMB)+"MB")))
+	for availableMB < diskWriteMB {
+		log.Warn(ctx, fmt.Sprintf("Insufficient available disk space of %vMB, reducing the test size of %vMB to %vMB. Note that this might result in lower performance", availableMB, diskWriteMB, diskWriteMB/2), nil)
+		diskWriteMB /= 2
+		if diskWriteMB == 0 {
+			return 0, "", errors.New("insufficient available disk space", z.Str("available_space", strconv.Itoa(availableMB)+"MB"))
 		}
 	}
 
-	var diskWriteTotal float64
-	for range diskWriteLoops {
-		time, err := writeFile(float64(actualDiskWriteMB) / 1024)
-		if err != nil {
-			return failedTestResult(testRes, err)
-		}
-		diskWriteTotal += time
+	testFilePath := dir + "/fioTest"
+
+	return diskWriteMB, testFilePath, nil
+}
+
+func performanceDiskWriteSpeedTest(ctx context.Context, conf *testPerformanceConfig) testResult {
```

Comment:
Wondering if we could run one `fio` test for read/write, and read disk speed and IOPS from the same output.

Reply:
Yes, that's definitely possible, we will save some time running the command. But on the other hand, I want to make the tests completely self sufficient and 1 test = 1 metric. Also easier to test 1 single thing by filtering the tests with the `--testcases` flag. EOD we are not wasting so much time running `fio`, compared to other performance tests here like the network connectivity.

Reply:
I've seen several examples of `fio` commands that have a runtime defined in order to have trusted results.

Maybe we need  to run `fio` with `--runtime=30s`?

Reply:
I already have `--runtime=60s` in my `fio` command. This runtime though is not really a runtime, but more of a... [timeout](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-runtime).

If we want to have continuous tests over X period of time, we can use probably the [time_based parameter](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-time_based).